doc/CONFIGURATION.md
# Configuration

Example:

```toml
[styles.haskell]
line = ["--"]
block-begin = ["{-"]
block-end = ["-}"]
block-nesting = true
strings = [
    { begin="\"", escape="\\\"", end="\"" },
    # Note: characters must also be treated as strings
    # Note: characters could be omitted here since all
    #       comment tokens are longer than 1 character
    #       and strings only exist to not wrongly match
    #       comment tokens.
    { begin="'", escape="\\'", end="'" },
]

[extensions]
haskell = ["hs"]

[aspell-options]
lang = "en_US"
ignore-case = "true"

[general]
highlight-commands = [
    "source-highlight -f esc --style-file=esc.style -i %FILE%",
    "pygmentize %FILE%",
]
italic-to-underline = true
box-style = "heavysharp"
# Ignore PascalCase, camelCalse, snake_case, and kebab-case words as
# they're likely identifiers.
filters = [
    "^[[:upper:]]?[[:lower:]]+[[:upper:]].*$",
    "^[[:alpha:]]*[_-][[:alpha:]].*$",
]
backup = false
```

## Sections

### `[styles]`

Contains the [comment style definitions](#comment-styles).

### `[extensions]`

Defines which file extension uses which comment style.
The keys should be one of the names defined in the `styles` section.

### `[aspell-options]`

Contains options that are forwarded to the Aspell library.
See `man spell` for available options.

Note that the values are always strings so for a boolean use `"true"` instead of `true`.

### `[general]`

Contains the main configuration for the program:

Key | Description | Default
---|---|---
`backup` | Whether to generate backup files | `true`
`bottom-status` | Whether to show the status bar at the bottom | `false`
`box-style` | Which flavor of box drawing characters to use, valid values are `"rounded"`, `"sharp"`, `"heavysharp"`, `"double"`, and `"ascii"`. An invalid value defaults to `rounded`. | `"rounded"`
`dim-code` | Whether to dim the colors of code outside comments | `true`
`filter-commented-code` | Whether filtering of commented code is enabled. More details about this are in the readme. | `false`
`filters` | A list of regular expressions, if any of them matches a word it is not checked. They use the RE2 syntax: https://golang.org/s/re2syntax (like Perl or Python). | `[]`
`highlight-commands` | A list of commands to try for highlighting.These should produce highlighting using ANSI escape codes. In the strings `%FILE%` is replaced with the filename. The first highlighter that does not give an error is used. | `[]`
`ignore-case` | Whether to ignore the case for the "Ignore all", "Replace all", and ignore lists. This does not affect the spell checking, use the `ignore-case` option in the `aspell-options` section for that. | `true`
`ignore-lists` | List of [ignore lists](#ignore-lists) | `[".spellcheck_comments_ignorelist"]`
`italic-to-underline` | Whether to convert the italic styles to underline in the highlighted source. This exists because some terminals don't support the italic style and treat it as reversed colors instead. | `false`
`layout` | The layout to use, either `"aspell"` or `"default"` (anything else defaults to `"default"`) | `"default"`
`mouse` | Whether to enable mouse interaction | `true`
`suggestions` | The maximum number of suggestions to show | `20` in default layout, `10` in Aspell layout
`tab-size` | Width of tab characters | `4`

### `[colors]`

Defines the interface colors, colors are given as ANSI escape codes:

key | description | default
---|---|---
`box-outline` | Border colors for box drawing characters | `"\x1b[38;5;213m"`
`comment` | Overwrites the comment color generated by highlighters. | If at least 1 highlighter is provided `""`, otherwise `"\x1b[1;32m"`. If all highlighters fail the latter is used as well.
`current-line-number` | The line number of the current line | `\x1b[38;5;251m`
`line-number` | All other line numbers | `\x1b[38;5;243m`
`menu` | The color for the suggestion menu in the default layout | `\x1b[48;5;61;38;5;232m`
`status-bar` | The colors for the status bar | `\x1b[38;5;251;7m`

## Comment styles

Note: tokens are limited to 8 bytes.

key | description
---|---
`line` | List of tokens that start a line comment
`block-begin` | List of tokens that start a block comment
`block-end` | List of tokens that end a block comment
`block-nesting` | Whether nesting of block comments is allowed
`strings` | List of string styles

The tokens in `block-begin` and `block-end` must match,
if for example the 2nd token in `block-begin` is matched only the 2nd token in `block-end` can terminate that comment.

Strings are used so we don't accidentally match a comment token inside a string.

The parser will match a token as soon as it can so if 2 tokens start with the same substring the longer of them can never be matched.
This means that strings cannot be matched in Python code as they would always steal away the token from doc-strings.

### Strings

key | description
---|---
`begin` | Token that begins a string
`end` | Token that ends a string
`escape` | Escape token that will not end the string.

There is no special support for character literals.

## Ignore lists

These are files containing 1 word per line, these words are never spell checked.
If a filename is absolute is it used as-is, if it's relative it may the relative
to the `spellcheck_comments.toml`/`config.toml` file and the current directory.
Case sensitivity of the words is controlled by the `general.ignore-case` option.


common/config.go
package common

import (
	"errors"
	"fmt"
	"os"
	"sort"
	"strings"

	"github.com/pelletier/go-toml/v2"
)

const (
	commentColorDefault = "\000DEFAULT"
	DefaultCommentColor = "\x1b[1;32m"
)

var FallbackCommentColor string

type CfgGeneral struct {
	Backup              bool     `toml:"backup"`
	BottomStatus        bool     `toml:"bottom-status"`
	BoxStyle            string   `toml:"box-style"`
	DimCode             bool     `toml:"dim-code"`
	FilterCommentedCode bool     `toml:"filter-commented-code"`
	Filters             []string `toml:"filters"`
	HighlightCommands   []string `toml:"highlight-commands"`
	IgnoreCase          bool     `toml:"ignore-case"`
	IgnoreLists         []string `toml:"ignore-lists"`
	ItalicToUnderline   bool     `toml:"italic-to-underline"`
	Layout              string   `toml:"layout"`
	Mouse               bool     `toml:"mouse"`
	Suggestions         int      `toml:"suggestions"`
	TabSize             int      `toml:"tab-size"`
}

type CfgColors struct {
	BoxOutline        string `toml:"box-outline"`
	Comment           string `toml:"comment-color"`
	CurrentLineNumber string `toml:"current-line-number"`
	LineNumber        string `toml:"line-number"`
	Menu              string `toml:"menu"`
	StatusBar         string `toml:"status-bar"`
}

type Config struct {
	Extensions    map[string][]string
	Styles        map[string]CommentStyle
	General       CfgGeneral
	Colors        CfgColors
	AspellOptions map[string]string `toml:"aspell-options"`
}

func DefaultConfig() Config {
	return Config{
		Extensions: make(map[string][]string),
		Styles:     make(map[string]CommentStyle),
		General: CfgGeneral{
			Backup:              true,
			BottomStatus:        false,
			BoxStyle:            "rounded",
			DimCode:             true,
			FilterCommentedCode: false,
			Filters:             []string{},
			HighlightCommands:   []string{},
			IgnoreCase:          true,
			IgnoreLists:         []string{".spellcheck_comments_ignorelist"},
			ItalicToUnderline:   false,
			Layout:              "default",
			Mouse:               true,
			Suggestions:         -1,
			TabSize:             4,
		},
		Colors: CfgColors{
			BoxOutline:        "\x1b[38;5;213m",
			Comment:           commentColorDefault,
			CurrentLineNumber: "\x1b[38;5;251m",
			LineNumber:        "\x1b[38;5;243m",
			Menu:              "\x1b[48;5;61;38;5;232m",
			StatusBar:         "\x1b[38;5;251;7m",
		},
		AspellOptions: make(map[string]string),
	}
}

func LoadConfig(pathname string) Config {
	data, _ := os.ReadFile(pathname)
	cfg := DefaultConfig()
	err := toml.Unmarshal(data, &cfg)
	var derr *toml.DecodeError
	if errors.As(err, &derr) {
		Fatal("%v:\n%s", err, derr.String())
	}
	for name, style := range cfg.Styles {
		if err := style.Check(); err != nil {
			Fatal("invalid comment style: %s: %s", name, err)
		}
	}
	if cfg.Colors.Comment == commentColorDefault {
		if len(cfg.General.HighlightCommands) == 0 {
			cfg.Colors.Comment = DefaultCommentColor
		} else {
			cfg.Colors.Comment = ""
		}
		FallbackCommentColor = DefaultCommentColor
	} else {
		FallbackCommentColor = cfg.Colors.Comment
	}
	if cfg.General.Suggestions < 0 {
		if cfg.General.Layout == "aspell" {
			cfg.General.Suggestions = 10
		} else {
			cfg.General.Suggestions = 20
		}
	}
	return cfg
}

func (self *Config) GetStyle(extension string) CommentStyle {
	for style, extensions := range self.Extensions {
		for _, ext := range extensions {
			if ext == extension {
				return self.Styles[style]
			}
		}
	}
	panic("unreachable")
}

func (self *Config) Aspell() map[string]string {
	return self.AspellOptions
}

func (self *Config) DumpStyles() {
	type Style struct {
		name  string
		style CommentStyle
	}
	styles := []Style{}
	for name, style := range self.Styles {
		styles = append(styles, Style{name, style})
	}
	sort.Slice(styles, func(i, j int) bool {
		a := styles[i].name
		b := styles[j].name
		aBuiltin := strings.HasPrefix(a, "builtin")
		bBuiltin := strings.HasPrefix(b, "builtin")
		if aBuiltin && !bBuiltin {
			return true
		} else if bBuiltin && !aBuiltin {
			return false
		}
		if aBuiltin {
			a = a[8:]
			b = b[8:]
		}
		return a < b
	})
	first := true
	for _, pair := range styles {
		if !first {
			fmt.Println()
		}
		first = false
		pair.style.Dump(pair.name, self.Extensions[pair.name])
	}
}
